### PR TITLE
AppsFeature now ensures that the Stratis apps folder exists where ena…

### PR DESCRIPTION
**Pull request by herbepau**
…bled through UseApps. @dangershony

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs [hunk 1]
```diff
@@ -56,12 +61,23 @@ public static IFullNodeBuilder UseApps(this IFullNodeBuilder fullNodeBuilder)
                     .AddFeature<AppsFeature>()
                     .FeatureServices(services =>
                     {
+                        TryEnsureAppsFolderExists(services);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs]: I don't think this is the right place to do this, it should go in the Feature `Initialize` method

File: src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs [hunk 2]
```diff
@@ -56,12 +61,23 @@ public static IFullNodeBuilder UseApps(this IFullNodeBuilder fullNodeBuilder)
...
+
                         services.AddSingleton<IAppsStore, AppsStore>();                        
                         services.AddSingleton<IAppsHost, AppsHost>();
                         services.AddSingleton<AppsController>();
                     });
             });
 
+            void TryEnsureAppsFolderExists(IServiceCollection services)
+            {
+                ServiceDescriptor descriptor = services.FirstOrDefault(x => x.ServiceType == typeof(DataFolder));
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs]: You dont need to query service just put DataFolder in the feature constructor and you will get it.

File: src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs [hunk 3]
```diff
@@ -56,12 +61,23 @@ public static IFullNodeBuilder UseApps(this IFullNodeBuilder fullNodeBuilder)
...
                     {
+                        TryEnsureAppsFolderExists(services);
+
                         services.AddSingleton<IAppsStore, AppsStore>();                        
                         services.AddSingleton<IAppsHost, AppsHost>();
                         services.AddSingleton<AppsController>();
                     });
             });
 
+            void TryEnsureAppsFolderExists(IServiceCollection services)
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs]: access modifier

**Comment by bokobza** [comment]
My suggestion is to get the DataFolder in the Feature's constructor, and then do this in the initialize method:
`Directory.CreateDirectory(this.dataFolder.ApplicationsPath);`

That's all you need.

**Comment by bokobza** [comment]
I think this will work:

```
private readonly DataFolder dataFolder;

public AppsFeature(ILoggerFactory loggerFactory, IAppsStore appsStore, IAppsHost appsHost, DataFolder dataFolder)
{
	this.appsStore = appsStore;
	this.appsHost = appsHost;
	this.dataFolder = dataFolder;
	this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
}

public override void Initialize()
{
	this.logger.LogInformation("Initializing {0}", nameof(AppsFeature));

	// Create the apps folder if it does not exist.
	Directory.CreateDirectory(this.dataFolder.ApplicationsPath);

	this.appsHost.Host(this.appsStore.Applications);
}
```

**Comment by herbepau** [comment]
ok

**Pull request merged**